### PR TITLE
Issue 43: Select Options - *BACKWARDS, *STARTKEY & *ENDWHERE & *ENDWHERESQL

### PR DESCRIPTION
**Pull request by Manish-Celestial**

### CODE REVIEW

**Review by robe070 — COMMENTED** [review]
Thats what I intended. Good work.
I've forced my branch back to before those 3 commits so I could pull your PR.

**Comment by Manish-Celestial** [comment]
Adding VerifierTestReport for windows platform
[Verifier_Test_Report.txt](https://github.com/lansa/db-regression/files/14060839/Verifier_Test_Report.txt)

File: Partition (WBP)/Functions/VI0043C.VTI0043.fun.yml [hunk 1]
```diff
@@ -308,7 +438,1635 @@
...
+    #Com_Owner.mthAdd_SubMsg( ER ('- Expected # rows ' + #LF_WorkingCount.asString) ('Actual rows ' + #LF_DB_Count.AsString) )
+    Return
+    Endif
+    
+    * Loop through the original list backwards and the DB list forwards (which has the BACKWARDS rows)
+    #LF_j := 1
+    #LF_From := (#LF_BIFN0700A_ToCount - #LF_BIFN0700A_FromCount + 1) * #LF_BIFC010A_COUNT
+    Begin_Loop Using(#LF_i) From(#LF_From) To(1) Step(-1)
+    Get_Entry Number(#LF_i) From_List(#w_lOriginal) Ret_Status(#this_rc)
+    #LF_BIFU100C := #BIFU100C
```
Comment (robe070) [inline Partition (WBP)/Functions/VI0043C.VTI0043.fun.yml:531]: Need to add a line here
```
#LF_BIFN0700A := #BIFN0700A
```

File: Partition (WBP)/Functions/VI0043C.VTI0043.fun.yml [hunk 2]
```diff
@@ -308,7 +438,1635 @@
...
+    Endif
+    
+    Get_Entry Number(#LF_j) From_List(#w_lDB) Ret_Status(#this_rc)
+    If (#this_rc <> OK)
+    #Com_Owner.mthIssue_Runtime_Message( #this_rc )
+    #Com_Owner.mthAdd_SubMsg( ER ('- Unexpected ' + #this_rc) ('Entry Number ' + #LF_j.AsString + ' may not exist in list LF_lDB') )
+    Leave
+    Endif
+    
+    If (#LF_BIFU100C <> #BIFU100C)
```
Comment (robe070) [inline Partition (WBP)/Functions/VI0043C.VTI0043.fun.yml]: For example this line in every test needs to be 
```
if ( #LF_BIFN0700A <> #BIFN0700A )
```

**Review by robe070 — CHANGES_REQUESTED** [review]
The tests are all wrong now because of the keys we changed.

All the tests are effectively finding all rows matching 2, say, and then checking if that key field = 2. Thats not testing anything.

The key is correct. Its BIFN0700A that needs to be tested for EVERY test. Originally the spec had:
_The test data is the same for all the secondary keys, matching BIFC010A._
You changed that secondary key to be BIFN0700A. Thus it is what needs to be compared.
I've commented on one test case as to what needs to change as an example for all the test cases.

**Review by robe070 — CHANGES_REQUESTED** [review]
You need *BACKWARDS  Descending key.

VI0049A seems to be out of date. It does not run VI0049D.

Please add log messages to display the values of the key returned by the SELECTs. Like you put in when we were in the meeting. I need to be clear that the tests are a good one.
Condition each of the tests with a global boolean so they can be easily disabled. E.g. #DetailedLog

**Comment by robe070** [comment]
And you need to run the test against IBMi. The results will be different - hence the conditional setting of the expected results for IBMi in V57033F.

**Comment by satishCodeIndians** [comment]
Submitted Changes, Once the windows platform changes are approved, I will check-in to IBMi and update the test as per the results over there

File: Partition (WBP)/Functions/VI0043D.VTI0043.fun.yml [hunk 3]
```diff
@@ -32,7 +32,7 @@
     Define Field(#LF_Total_Warnings) Reffld(#Std_Count)
     Define Field(#LF_Execution_Type) Reffld(#Std_Alpha)
     Define Field(#LF_Remote_Platform) Reffld(#BIFA001A)
-    
+    Define Field(#LF_DetailedLog) Reffld(#Std_Count)
```
Comment (robe070) [inline Partition (WBP)/Functions/VI0043D.VTI0043.fun.yml]: I said a boolean
```
Define Field(#LF_DetailedLog) Reffld(#BIFB001B)
```

File: Partition (WBP)/Functions/VI0043D.VTI0043.fun.yml [hunk 4]
```diff
@@ -326,6 +353,11 @@
     Select Fields(#BIFINT2S) From_File(LI0043DI2) Where(((#BIFINT2S *LE 111) *And (#BIFN0700A *LE 190)) *Or ((#BIFINT2S *GE -111) *And (#BIFN0700A *LE 510))) With_Key(#this_test_value) Nbr_Keys(*COMPUTE) Generic(*YES) Io_Status(#this_rc) Io_Error(*NEXT) Val_Error(*NEXT) Options(*ENDWHERE)
     #LF_Func_LI0043DI2 += 1
     #Com_Owner.mthAdd_TraceMsg( #this_rc '' ('Result: ' + #BIFINT2S.asString) )
+    
+    * Log the Key values selected to the Verifier Log
+    If (#LF_DetailedLog <> 0)
```
Comment (robe070) [inline Partition (WBP)/Functions/VI0043D.VTI0043.fun.yml]: Change all these tests to
```
If (#LF_DetailedLog)
```

**Review by robe070 — CHANGES_REQUESTED** [review]
You need to output the WHERE clause to the log file too.

**Comment by robe070** [comment]
Next time provide me with a detailed log and I can reply sooner.

**Comment by robe070** [comment]
The WHERE clause needs to be changed. The very first result set should stop at 111. It doesn't. It goes to 119.
(#BIFINT2S *LE 111) *And (#BIFN0700A *LE 190)) *Or ((#BIFINT2S *GE -111) *And (#BIFN0700A *LE 510))

**Comment by satishCodeIndians** [comment]
[Verifier_Test_Report_I0043D_2.txt](https://github.com/lansa/db-regression/files/14121825/Verifier_Test_Report_I0043D_2.txt)

**Comment by satishCodeIndians** [comment]
Super Server tests passing
[Verifier_Test_Report_I0043D_3.txt](https://github.com/lansa/db-regression/files/14127858/Verifier_Test_Report_I0043D_3.txt)

File: Partition (WBP)/Functions/VI0043D.VTI0043.fun.yml [hunk 5]
```diff
@@ -332,6 +332,11 @@
     Define_Map For(*INPUT) Class(#BIFINT2S) Name(#this_test_value)
     Define_Map For(*BOTH) Class(#BIFA002A) Name(#this_rc)
     
+    Define Field(#LF_Where_Clause) Reffld(#BIFS100A)
+    #LF_Where_Clause := 'Where(((#BIFINT2S *LE 111) *And (#this_test_value *GE 0)) *Or ((#BIFINT2S *GE -111) *And (#this_test_value *LE 0)))'
```
Comment (robe070) [inline Partition (WBP)/Functions/VI0043D.VTI0043.fun.yml]: Its being truncated. Split this in 2
```
#LF_Where_Clause := 'Where(((#BIFINT2S *LE 111) *And (#this_test_value *GE 0)) *Or'
#LF_Where_Clause2 := '((#BIFINT2S *GE -111) *And (#this_test_value *LE 0)))'
#Com_Owner.mthAdd_SubMsg( #this_rc (#this_operation + ' ' + #LF_Where_Clause) #LF_Where_Clause2 )
```

File: Partition (WBP)/Functions/VI0043D.VTI0043.fun.yml [hunk 6]
```diff
@@ -350,13 +355,14 @@
...
-    Select Fields(#BIFINT2S) From_File(LI0043DI2) Where(((#BIFINT2S *LE 111) *And (#BIFN0700A *LE 190)) *Or ((#BIFINT2S *GE -111) *And (#BIFN0700A *LE 510))) With_Key(#this_test_value) Nbr_Keys(*COMPUTE) Generic(*YES) Io_Status(#this_rc) Io_Error(*NEXT) Val_Error(*NEXT) Options(*ENDWHERE)
+    Select Fields(#BIFINT2S) From_File(LI0043DI2) Where(((#BIFINT2S *LE 111) *And (#this_test_value *GE 0)) *Or ((#BIFINT2S *GE -111) *And (#this_test_value *LE 0))) With_Key(#this_test_value) Nbr_Keys(*COMPUTE) Generic(*YES) Io_Status(#this_rc) Io_Error(*NEXT) Val_Error(*NEXT) Options(*ENDWHERE)
     #LF_Func_LI0043DI2 += 1
     #Com_Owner.mthAdd_TraceMsg( #this_rc '' ('Result: ' + #BIFINT2S.asString) )
     
     * Log the Key values selected to the Verifier Log
-    If (#LF_DetailedLog <> 0)
+    If (#LF_DetailedLog)
     #Com_Owner.mthAdd_SubMsg( OK ' + Selected Data' (#BIFINT2S.AsString) )
+    #Com_Owner.mthAdd_SubMsg( OK (' Evaluated Condition : Where (' + (#BIFINT2S.AsString) + ' < 111 and' + #this_test_value.AsString + ' > 0 ) or ( ' + (#BIFINT2S.AsString) + ' > -111 and ' + #this_test_value.AsString + ' < 0 )') '' )
```
Comment (robe070) [inline Partition (WBP)/Functions/VI0043D.VTI0043.fun.yml]: Its being truncated. Split the WHERE clause across the two params of mthAdd_SubMsg
```
#Com_Owner.mthAdd_SubMsg( OK (' Evaluated Condition : Where (' + (#BIFINT2S.AsString) + ' < 111 and' + #this_test_value.AsString + ' > 0 ) or ') '( ' + (#BIFINT2S.AsString) + ' > -111 and ' + #this_test_value.AsString + ' < 0 )' )
```

**Review by robe070 — CHANGES_REQUESTED** [review]
I need the changes in order to be able to review that the tests are useful.
A quick turnaround would be useful.

**Comment by Manish-Celestial** [comment]
Sure, I Pushed the changes and here is the log
[Verifier_Test_Report_I0043_4.txt](https://github.com/lansa/db-regression/files/14132310/Verifier_Test_Report_I0043_4.txt)

**Review by robe070 — CHANGES_REQUESTED** [review]
Test values are all good.

`*BACKWARDS DESCENDING` has message `*FORWARDS DESCENDING`

Minor issue. add a space here `and-100`:
```
Evaluated Condition : Where (-100 < 111 and-100 > 0 ) or ( -100 > -111 and -100 < 0 )
```

**Comment by Manish-Celestial** [comment]
Done. Thank you

**Comment by robe070** [comment]
The IBMi test is still failing.

**Comment by Manish-Celestial** [comment]
[Verifier_Test_Report_I0043_5.txt](https://github.com/lansa/db-regression/files/14138578/Verifier_Test_Report_I0043_5.txt)

**Comment by Manish-Celestial** [comment]
The tests are successful on IBMi with latest changes. Attaching the report

**Pull request merged**